### PR TITLE
HMAN-98 Add missing list assist fields. Refactor and correct unit tests.

### PR DESCRIPTION
**Pull request by markdathornehmcts**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/HMAN-98

### Change description ###

Add missing List Assist fields in the POST/PUT Request payload

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/hmc/controllers/HearingManagementController.java [hunk 1]
```diff
@@ -130,10 +132,31 @@ public GetHearingsResponse getHearings(@PathVariable("ccdCaseRef")
...
+        );
+    }
+
+    @PutMapping(path = "/test/{id}", consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
+    @ResponseStatus(HttpStatus.CREATED)
+    public HmiSubmitHearingRequest test(@RequestBody UpdateHearingRequest hearingRequest,
+                                        @PathVariable("id") Long hearingId) {
+        hearingManagementService.updateHearingRequest(hearingId, hearingRequest);
+        return hearingManagementService.test(hearingId, hearingRequest);
+    }
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/controllers/HearingManagementController.java:158]: Can we delete these temporary test methods? I don't think we need them anymore since we now have the functionality in place to put messages on the queue. We can test them properly now.

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/HmiCaseDetailsMapper.java [hunk 2]
```diff
@@ -26,6 +31,38 @@ public HmiCaseDetails getCaseDetails(CaseDetails caseDetails, Long hearingId) {
             .caseRegistered(caseDetails.getCaseSlaStartDate())
             .caseInterpreterRequiredFlag(caseDetails.getCaseInterpreterRequiredFlag())
             .caseRestrictedFlag(caseDetails.getCaseRestrictedFlag())
+            .caseLinks(getCaseLinksArray(Arrays.asList(caseDetails.getCaseDeepLink())))
+            .casePublishedName(caseDetails.getPublicCaseName())
+            .caseAdditionalSecurityFlag(caseDetails.getCaseAdditionalSecurityFlag())
             .build();
     }
+
+    public HmiCaseDetails getCaseDetails(CaseDetails caseDetails, Integer versionNumber, Long hearingId) {
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/HmiCaseDetailsMapper.java:40]: Looks like the only difference is the addition of the `versionNumber` and rest of the builder is the same?
I think we should only need one `getCaseDetails()` method so can delete the previous one? Every request will have a version number in the `case_hearing_request` table.

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/HmiSubmitHearingRequestMapper.java [hunk 3]
```diff
@@ -36,4 +38,22 @@ public HmiSubmitHearingRequest mapRequest(Long hearingId, HearingRequest hearing
             .build();
     }
 
+    public HmiSubmitHearingRequest mapRequest(Long hearingId, UpdateHearingRequest hearingRequest) {
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/HmiSubmitHearingRequestMapper.java:41]: Should be able to remove this method completely and avoid the duplication if we have `versionNumber` as an argument in the existing `mapRequest()`.

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/ListingOtherConsiderationsMapper.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+public class ListingOtherConsiderationsMapper {
+
+    public List<String> getListingOtherConsiderations(Boolean hearingInWelshFlag,
+                                                      List<String> facilityTypes) {
+        List<String> listOtherConsiderations = new ArrayList<>();
+        if (null != hearingInWelshFlag) {
+            listOtherConsiderations.add(hearingInWelshFlag.toString());
+        } else {
+            listOtherConsiderations.add(Boolean.FALSE.toString());
+        }
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/ListingOtherConsiderationsMapper.java:18]: I don't think the ticket is clear enough about this one, but this logic doesn't sound right.

On our scope doc it says that we should be setting it to some system defined variable when `hearingInWelshFlag` is true. I don't think we should be setting it simply to the string value "true"! And when false it doesn't sound like we should be putting anything.

Could you catch up with the BAs on this please for clarification? The scope doc still has the value we should set it to as undefined!

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/ListingOtherConsiderationsMapper.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        if (null != facilityTypes) {
+            facilityTypes.stream().forEach(listOtherConsiderations::add);
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/ListingOtherConsiderationsMapper.java:20]:
```suggestion
            listOtherConsiderations.addAll(facilityTypes);
```

File: src/main/java/uk/gov/hmcts/reform/hmc/model/hmi/HmiCaseDetails.java [hunk 6]
```diff
@@ -34,4 +34,12 @@
 
     private Boolean caseRestrictedFlag;
 
+    private Integer caseVersionId;
+
+    private CaseLinks[] caseLinks;
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/model/hmi/HmiCaseDetails.java:39]:
```suggestion
    private List<CaseLinks> caseLinks;
```
Might save some of the converting between lists and arrays that is done elsewhere.

File: src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementService.java [hunk 7]
```diff
@@ -25,8 +25,13 @@
 
     void sendResponse(String json);
 
-    void sendRequestToHmiAndQueue(Long hearingId, HearingRequest hearingRequest, String messageType);
+    void sendRequestToHmiAndQueue(CreateHearingRequest hearingRequest, Long hearingId, String messageType);
+
+    void sendRequestToHmiAndQueue(UpdateHearingRequest hearingRequest, Long hearingId, String messageType);
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementService.java:30]: Hopefully based on the other comments we can undo these changes to keep it more re-usable

**Review by danlysiak — COMMENTED** [review]
A few suggestions to keep things cleaner and avoid duplication. One comment that needs BA input.

File: src/main/java/uk/gov/hmcts/reform/hmc/controllers/HearingManagementController.java (hunk 1 above)
Comment (markdathornehmcts) [reply src/main/java/uk/gov/hmcts/reform/hmc/controllers/HearingManagementController.java]: Yup. Done. Padma showed me how to use the queues instead.

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/ListingOtherConsiderationsMapper.java (hunk 4 above)
Comment (markdathornehmcts) [reply src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/ListingOtherConsiderationsMapper.java]: Query with Arup atm.

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/ListingOtherConsiderationsMapper.java (hunk 5 above)
Comment (markdathornehmcts) [reply src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/ListingOtherConsiderationsMapper.java]: Done

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/ListingOtherConsiderationsMapper.java (hunk 4 above)
Comment (markdathornehmcts) [reply src/main/java/uk/gov/hmcts/reform/hmc/helper/hmi/ListingOtherConsiderationsMapper.java]: "CY" value applied for now as advised.

File: src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementService.java [hunk 8]
```diff
@@ -25,8 +25,13 @@
...
 
     void sendRequestToHmiAndQueue(DeleteHearingRequest hearingRequest,Long hearingId, String messageType);
 
+    HmiSubmitHearingRequest test(Long hearingId, CreateHearingRequest hearingRequest);
+
+    HmiSubmitHearingRequest test(Long hearingId, UpdateHearingRequest hearingRequest);
```
[the remark this answers is not in the document]
Comment (markdathornehmcts) [reply src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementService.java:35]: Gone
